### ISSUE TITLE
[REVIEW]: HRDS: A Python package for hierarchical raster datasets

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1112]
**Submitting author:** @jhill1 (<a href="http://orcid.org/0000-0003-1340-4373">Jon Hill</a>)
**Repository:** <a href="https://github.com/EnvModellingGroup/hdrs" target ="_blank">https://github.com/EnvModellingGroup/hdrs</a>
**Version:** 0.1.2
**Editor:** @kthyng
**Reviewer:** @edoddridge, @PythonCHB
**Archive:** <a href="https://doi.org/10.5281/zenodo.2691685" target="_blank">10.5281/zenodo.2691685</a>

## Status

[![status](http://joss.theoj.org/papers/d8f85d9247eef870a36ce3a9ccb6b64b/status.svg)](http://joss.theoj.org/papers/d8f85d9247eef870a36ce3a9ccb6b64b)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/d8f85d9247eef870a36ce3a9ccb6b64b"><img src="http://joss.theoj.org/papers/d8f85d9247eef870a36ce3a9ccb6b64b/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/d8f85d9247eef870a36ce3a9ccb6b64b/status.svg)](http://joss.theoj.org/papers/d8f85d9247eef870a36ce3a9ccb6b64b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/EnvModellingGroup/hdrs" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@edoddridge & @ PythonCHB, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @kthyng know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/EnvModellingGroup/hdrs">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** 0.1.2
- [x] **Authorship:** Has the submitting author (@jhill1) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/EnvModellingGroup/hdrs">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** 0.1.2
- [x] **Authorship:** Has the submitting author (@jhill1) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @edoddridge, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01112/joss.01112/10.21105.joss.01112.pdf)

**Comment by edoddridge** [comment]
I'm happy to say that I've finished my review of `HRDS` and can recommend it for publication. 👍

It is a well-written piece of code that solves a clearly stated problem. 'HRDS' provides a clean interface to mesh together multiple files with different resolutions and create manageable input files for variable resolution geophysical modelling.

While I've opened a number of issues in the code repository, they are all quite minor, and none of them should prevent publication of the paper. With perhaps one exception - it would be a shame to publish the paper before the typos are fixed. 

I had very little difficulty obtaining a working version of the code. The dependencies were clearly identified. I encountered only one problem, and the documentation has now been updated to guide users around that issue. The test suite worked straight out of the box. 

My only disappointment regarding the testing is the lack of a code coverage tool. The `.travis.yml` file contains some commented out lines for using coveralls to assess code coverage. I would recommend (re)implementing a code coverage tool.

The python code is laid out in a clear and logical way. I found it quite easy to work out what each piece of code does.

My biggest concern with this submission is the lack of a real-life worked example. However, @jhill1 has obtained the necessary data and has [committed to implementing one soon](https://github.com/EnvModellingGroup/hdrs/issues/13). Once the data is included, the test suite could be altered to use those datasets instead of the proprietary ones that @jhill1 uses locally.

The overarching documentation is currently limited to the readme. However, given the nature of the software, this is perfectly acceptable. API documentation is contained in docstrings for the various methods and classes. Having extensive docstrings means that Sphinx based documentation could be added in the future without too much extra effort.

**Comment by jhill1** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01112/joss.01112/10.21105.joss.01112.pdf)

**Comment by PythonCHB** [comment]
# Review of the HRDS software package:

https://github.com/EnvModellingGroup/hrds

Overall, this looks like a small, but useful package for raster data processing. But there are some issues with the pacakge structure, tests, and documentation, outlines below.

I recommend that the package be published once the below issues are addressed.

## Installation, etc:

### Platform support:

From the README:

"These instructions assume a Debian-based Linux. HDRS should work on other
systems, but is currently untested."

There are a lot of users out there that are not running Debian-based Linux -- this project should attempt to provide *some* guidance for other platforms. UNfotunately, the GDAL stack is pretty ugly to install, so perhaps recommending conda and conda-forge would be a good way to get people rolling in a platform independent way.

I notice that conda is (partially) being used for the CI -- so could full conda instuctions could be provided.

I am testing on OS-X with conda and conda-forge, Python 3.7

In fact, I highly recommend that the project authors make hrds available in conda-forge: https://conda-forge.org/#add_recipe

## Documentation and Examples:

In the main README, after providing Debian-only install instructions, the primary example is using the `firedrake` and `thetis` packages -- nifty use-case, but it would be much better, for a more general purpose package to have the primary examples be based only of the core package and its dependencies.

### Install

The hrds package is using a standard setuptools / setup.py install. However, there are some issues:

The tests are put in a package that is next to the hrds package, and the setup.py uses setuptools.find_packages to to identify the packages to install. So when run, you get a top-level package called "tests" with the hrds tests in it. This is not good.

Here are some thoughts on where to put tests:

http://pythonchb.github.io/PythonTopics/where_to_put_tests.html

I've provided a PR here that puts the tests internal to the package:

https://github.com/EnvModellingGroup/hrds/pull/18

I also added formatting, etc changes to the setup.py in that PR.

### tests:

* The way the test files were found was fragile -- so would only run if the current working dir was the parent of the tests dir:

https://github.com/EnvModellingGroup/hrds/issues/21

* There is a duplicate test name in: test_hrds.py -- those are skipped anyway due to lack of GEBCO data, but that should be fixed! Issue here:

https://github.com/EnvModellingGroup/hrds/issues/20

PR for fixes to the file finding, and PEP8 compliance fixes is here:

https://github.com/EnvModellingGroup/hrds/pull/24

### temp files.

The `HRDS` class appears to be creating temp buffer files that it does not delete:

https://github.com/EnvModellingGroup/hrds/issues/23

NOTE: computer memory is pretty huge these days -- are the buffer files even needed -- maybe only used when the rasters are really big?

## JOSS paper

The JOSS paper is fine as it goes, but there are a few issues:

* There are a few references cited, but nothing in the references section -- are those defined somewhere else?

* "a linear distance buffer is created to smoothly blend the two datasets together."
  
  I would like to see some description (and ideally a reference) for this algorithm.

* The docs do note that two rasters "are unlikely to 
agree on the topographic/bathymetric height where they overlap", but there is no mention of (or algorithms for?) accounting for major differences -- different vertical datums, for instance.

## Documentation

The only user docs I could find is the README and docstrings. Some more comprehensive user and reference docs would be good. Also, the docstrings are not complete nor PEP 257 compliant.

**Comment by kthyng** [comment]
Great! Thanks so much for the reviews, @edoddridge and @PythonCHB. 

It looks like there are some unchecked check boxes for both reviews currently, and @jhill1 can work on addressing issues that have been brought up. Thanks again everyone.

**Comment by kthyng** [comment]
Just a note that I will have an away message popping up starting tomorrow and I will continue to respond here while I can; once I can't and if this review is still on-going, then another editor will take over.

**Comment by jhill1** [comment]
Thanks both for the reviews (especially whilst furloughed @PythonCHB - much appreciated).

**Comment by lheagy** [comment]
:wave: Hi @jhill1, @edoddridge, @PythonCHB: just a quick note to say that I am happy to answer any questions or help out on the editor side while @kthyng is away.

**Comment by arfon** [comment]
@jhill1 - I think we're waiting on your responses to @PythonCHB's review at this point.

**Comment by jhill1** [comment]
You are indeed. Swamped with teaching this term, but will get to this ASAP. Thanks for your patience!

**Comment by labarba** [comment]
I'm going to go ahead and add the "paused" label here. Do ping us when you are ready to re-engage with this submission, @jhill1.

**Comment by jhill1** [comment]
@whedon generate pdf

**Comment by kthyng** [comment]
Great @jhill1! 

@PythonCHB can you look through @jhill1's work to see if you are satisfied to check off the remaining boxes in your review, or if you have follow up comments?

**Comment by PythonCHB** [comment]
Thanks for addressing my concerns -- this is a more robust package now!

A few issues still:

## The Readme:

I have copy edited the README -- PR here:

https://github.com/EnvModellingGroup/hrds/pull/29

Most of that is just copy editing, but a small issue -- in the example in the readme, there is this code:

```
import sys
sys.path.insert(0,"../../")

from hrds import hrds
```

messing around with `sys.path` like that is not considered good practice these days. As the previous instructions are how to install the package, this is totally unnecessary. And if a user wants to work with the package directly from the source, the recommended way to do that is an "editable install":

`pip install -e ./`

and then the package will be available with a regular import.

I have removed that code in my PR.

In the `thetis` example, there is:

```
from firedrake import *
from thetis import *
from firedrake import Expression
```
`import *` is highly discouraged these days -- and particularly in example code like this, it's then confusing be3cause you don't know if a name comes from firedrake or thetis. And if you have imported everything from firedrake, you don't need to import Expression.

`import *` should be removed -- I have not done this in my PR, as I haven't installed thetis and firedrake to test.

## Documentation

I've added a few copy-edits tot he PR above.

Note that there is a fair bit of overlap between the README and the Sphinx Docs -- (and teh Sphinx docs now need copy editing to match the README). I suggest that you single source these if you can.

gitHub will allow READMEs in RST as well, so you could have one RST page with the intro and all,  and include that into your Sphinx docs.

You could also make the README smaller, and link to the more substantial docs -- which you could publish as html via gitHubIO or readthedocs.

### docstrings and the API docs:

It looks like Sphinx auto-doc is not including teh class __init__ docs -- which can be very helpful.

Also, some of those __init__ docstrings (hrds.hrds.HRDS.__init__, for instance, are not complete -- you should document all the parameters to the __init__.


## minor notes:

`from __future__ import print_function` is necessary for py3 only code.

No need to derive from `object` in py3 either.


### Code style:

I see this:

```
    def set_mask(self, mask):
        """
        Set a mask to use. Not yet implemented.
        """
        self.mask = mask
```

This is why Python has "properties" -- there is not need to write do-nothing getters and setters.

https://uwpce-pythoncert.github.io/PythonCertDevel/modules/Properties.html?highlight=properties

## The Paper

The References section is blank -- but there is at least one reference in the paper -- am I missing something?

I'd still like to see a more description the buffering algorithm -- linear interpolation does not need defining, but exactly how one interpolates between two different resolutions of rasters in the buffer zone is not obvious.

# Conclusion:

None of these issues are fatal -- congrats! You've made a nice contribution to the community.

**Comment by kthyng** [comment]
Thanks so much @PythonCHB! 

@jhill1 Can you address these comments soon?

**Comment by jhill1** [comment]
Thanks @PythonCHB !

The pull request has been merged (thanks for that!). 

**Documents**: I've pulled over the changes in the README to the Sphinx docs and will have a think about how to make the best of them. I have a summer student this year who will be using this code, so we'll see how they get on and then I might ask them to rationalise the docs along the lines proposed.

I've removed all the `import *` from docs and code examples and the leftover Python 2 code. I haven't removed the set_mask yet (as a reminder for me to implement this) - the code will have to do more than set the self.mask variable (e.g. regenerate buffers). 

**Paper:** I've added a figure to explain the buffer which will hopefully help The references are added automatically by JOSS when the PDF is created.

**Comment by jhill1** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01112/joss.01112/10.21105.joss.01112.pdf)

**Comment by jhill1** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01112/joss.01112/10.21105.joss.01112.pdf)

**Comment by kthyng** [comment]
Great work @jhill1 it looks like you have addressed all of @PythonCHB's comments. There is one unchecked box, but the pdf does indeed now have references included and they have doi's when appropriate.

I am going through my own checks now.

**Comment by kthyng** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1016/j.renene.2017.09.058 is OK
- 10.1016/j.renene.2014.11.079 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
@jhill1 your citation of Martin-Short2015 shouldn't have parentheses so that it is inline as it is part of the sentence as written.

**Comment by kthyng** [comment]
@jhill1 After that fix, please make a Zenodo archive, and report the DOI in this thread.

**Comment by jhill1** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01112/joss.01112/10.21105.joss.01112.pdf)

**Comment by jhill1** [comment]
Hi @kthyng. Thanks for editing this submission - it's been fun. Zenodo DOI: 10.5281/zenodo.2691685

**Comment by kthyng** [comment]
@jhill1 agreed!

Ok I just looked up [your submission to Zenodo](https://zenodo.org/record/2691685#.XNSRrpNKgWo) and the metadata includes @PythonCHB which is nice but not correct — this often happens. Can you edit the metadata so that it matches your paper submission? In particular you want the author list to match, and the title to match.

**Comment by jhill1** [comment]
Fixed. Will add the .json to fix that properly at some point. Thanks.

**Comment by kthyng** [comment]
Great!

**Comment by kthyng** [comment]
@whedon set 10.5281/zenodo.2691685 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.2691685" target="_blank">10.5281/zenodo.2691685</a> is the archive.

**Comment by kthyng** [comment]
@whedon set <v0.1.2> as version

**Comment by whedon** [comment]
OK. <v0.1.2> is the version.

**Comment by kthyng** [comment]
@whedon set 0.1.2 as version

**Comment by whedon** [comment]
OK. 0.1.2 is the version.

**Comment by kthyng** [comment]
Ok @openjournals/joss-eics this submission is ready to be officially accepted!!! Congrats @jhill1!!

**Comment by PythonCHB** [comment]
Not really a big deal: but if you are publishing this, maybe it's time for version 1.0?

If it's not really ready for 1.0, it's not ready for publication, either.

**Comment by labarba** [comment]
@jhill1 — would you like to up you version number? We don't require it: it's up to you.

**Comment by jhill1** [comment]
Leave for now, please. @PythonCHB is right that it should have been made v1.0, but I've already pushed pip and conda updates with 0.12 and referred to the version being the one published. I'd rather keep a cleaner history.

**Comment by labarba** [comment]
@PythonCHB — you have the "References" item unchecked. Did you find something missing there?

**Comment by labarba** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1016/j.renene.2017.09.058 is OK
- 10.1016/j.renene.2014.11.079 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by labarba** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1016/j.renene.2017.09.058 is OK
- 10.1016/j.renene.2014.11.079 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/689

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/689, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by labarba** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/690
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01112](https://doi.org/10.21105/joss.01112)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by labarba** [comment]
Congratulations, @jhill1, your JOSS paper is published 🚀 

Sincere thanks to our editor: @kthyng, and the reviewers: @edoddridge, @PythonCHB — we appreciate your enormous contribution! 🙏

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01112/status.svg)](https://doi.org/10.21105/joss.01112)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01112">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01112/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01112/status.svg
   :target: https://doi.org/10.21105/joss.01112
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01112/status.svg)](https://doi.org/10.21105/joss.01112)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss
